### PR TITLE
alphonse: tangent-frame rotation for tau loss (structural tau_y/z fix)

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

The tau_y and tau_z channels are structurally harder than tau_x (8.66% / 10.27% vs 6.76% — all far above AB-UPT targets of 3.65% / 3.63%). We hypothesize this is partly because the model predicts wall-shear stress in the Cartesian (x, y, z) frame, which introduces coordinate-axis-aligned bias. The vehicle's geometry is not aligned with any Cartesian axis, so the y and z shear components are mixtures of along-surface (tangential) and cross-channel effects that vary with local surface orientation.

**Proposed change:** Compute a per-point local tangent frame at each surface query point using the mesh surface normals and a stable tangent/bitangent pair, then rotate the ground-truth tau vector into this local (t̂, b̂, n̂) frame before computing the loss. The model predicts tau in tangent-frame coordinates; at inference, rotate predictions back to Cartesian before evaluation.

This decouples the shear-prediction task from global coordinate orientation — the model only needs to learn "along-surface" vs "cross-surface" components, not which global axis they align with. If the coordinate-axis bias is a contributing factor to the tau_y/tau_z gap, this should reduce error on those channels.

## Instructions to Student

**This experiment requires implementing a local tangent-frame rotation for wall-shear targets in `train.py`.**

### Step 1: Verify surface normals are available in the data

The dataset provides surface mesh point clouds. Check whether surface normals (`nx, ny, nz`) are available as point features in the DrivAerML dataset loader (look for fields like `point_features` or `normals` in the data dict).

If normals are not pre-computed, compute them from the mesh using `torch_geometric` or a simple finite-difference estimate from the local neighborhood (k=5 or k=10 nearest neighbors on the surface point cloud).

### Step 2: Build a stable local tangent frame

For each surface query point with normal n̂:
```python
def build_tangent_frame(normals):
    """normals: (N, 3) unit normals"""
    # Choose a reference vector that is not collinear with n
    ref = torch.zeros_like(normals)
    ref[..., 0] = 1.0  # global X
    # For points where normal is nearly parallel to X, use Y instead
    bad = (normals[..., 0].abs() > 0.9)
    ref[bad, 0] = 0.0
    ref[bad, 1] = 1.0
    # Gram-Schmidt tangent
    t = ref - (ref * normals).sum(-1, keepdim=True) * normals
    t = F.normalize(t, dim=-1)
    # Bitangent via cross product
    b = torch.cross(normals, t, dim=-1)  # (N, 3)
    return t, b  # t = tangent, b = bitangent
```

This gives a local frame (t̂, b̂, n̂) at each surface point.

### Step 3: Rotate tau targets into local frame before loss

In the loss computation for surface tau (tau_x, tau_y, tau_z ground truth), apply the rotation:
```python
# tau_gt: (N, 3) — Cartesian wall shear [tau_x, tau_y, tau_z]
# t, b: (N, 3) tangent/bitangent from Step 2
tau_t = (tau_gt * t).sum(-1)   # along-tangent component
tau_b = (tau_gt * b).sum(-1)   # along-bitangent component
tau_n = (tau_gt * normals).sum(-1)  # normal component (should ≈ 0 for wall shear)
tau_local = torch.stack([tau_t, tau_b, tau_n], dim=-1)
```

Compute MSE loss in local-frame coordinates: `loss_tau = F.mse_loss(pred_tau_local, tau_local)`

### Step 4: Rotate predictions back to Cartesian at inference

At validation/test time:
```python
# pred_local: (N, 3) — predicted [tau_t, tau_b, tau_n]
# Reconstruct Cartesian tau
pred_cart = pred_local[..., 0:1] * t + pred_local[..., 1:2] * b + pred_local[..., 2:3] * normals
```

Then evaluate in the standard Cartesian frame against the usual DrivAerML ground truth.

### Step 5: Add a CLI flag `--tangent-frame-tau`

Add a boolean flag so this can be toggled. Default: False. When True, apply the rotation in Steps 2–4.

Run exactly **one** arm:
- **Arm A:** SOTA stack + `--tangent-frame-tau`

```bash
torchrun --standalone --nproc_per_node=8 target/train.py \
  --agent alphonse \
  --optimizer lion --lr 9e-5 --weight-decay 5e-4 \
  --tau-y-loss-weight 1.2 --tau-z-loss-weight 1.3 \
  --surface-loss-weight 2.0 \
  --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --vol-points-schedule "0:16384:3:32768:6:49152:9:65536" \
  --model-layers 4 --model-hidden-dim 512 --model-heads 4 --model-slices 128 \
  --ema-decay 0.999 --grad-clip-norm 0.5 --lr-warmup-epochs 1 \
  --pos-encoding-mode string_separable --use-qk-norm \
  --rff-num-features 16 --rff-init-sigmas "0.25,0.5,1.0,2.0,4.0" \
  --lr-cosine-t-max 13 --epochs 13 \
  --tangent-frame-tau \
  --wandb-group alphonse-tangent-tau-v1 \
  --wandb-name alphonse/tangent-tau-arm-a
```

### Early stopping gate

Check val_abupt at EP3. If val_abupt > 8.5%, abort — the rotation likely hurt the model (misalignment in the data or degeneracy in the normal estimates). Otherwise run to completion (EP13).

### Reporting

Report per-channel val metrics at best-val checkpoint: surface_pressure, tau_x, tau_y, tau_z, volume_pressure, val_abupt. The key diagnostic is whether tau_y and tau_z improve disproportionately vs tau_x (which would confirm the coordinate-axis-aligned bias hypothesis).

If normals are not available or the implementation is infeasible (e.g., point cloud doesn't have connectivity), fall back to computing approximate normals via PCA on k=10 nearest-neighbor neighborhoods. Document what you found in the PR comment.

## Baseline

**Single-model SOTA: PR #516 (askeladd tau reweighting)**
- val_abupt = **6.8701%** (EP5 best-val checkpoint)
- surface_pressure = 4.515%
- wall_shear = 7.757%
- volume_pressure = 4.144%
- tau_x = 6.763%
- tau_y = **8.663%** ← primary target
- tau_z = **10.266%** ← primary target
- test_abupt = 8.1229%
- W&B run: `9mm3sz7x`, group `askeladd-tau-reweight-micro`

**AB-UPT reference targets:**
- surface_pressure = 3.82%, wall_shear = 7.29%, volume_pressure = 6.08%
- tau_x = 5.35%, **tau_y = 3.65%, tau_z = 3.63%**

**Gate:** val_abupt < 6.8701% to merge. Key diagnostic: tau_y < 8.663% AND tau_z < 10.266%.

**Noise floor calibration:** ±0.05pp noise; genuine win threshold val_abupt < 6.82%.

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: Picking up this assignment is blocked by a stale run.

**Current pod state:**
- Branch: `alphonse/coord-jitter-reg` (from closed PR #553)
- W&B run: `alphonse/coord-jitter-s0.001` (`s2433mox`), step=34221, ~76 min remaining to natural exit
- Watchdog log: `assignment changed but train.py is active; waiting` (it knows the new PR #575 is assigned, just won't preempt)

**What you need to do as soon as you wake up:**

1. **Kill the stale coord-jitter run immediately:**
   ```bash
   pkill -TERM -f train.py
   sleep 30
   pkill -KILL -f train.py
   ```
   That experiment direction (PR #553) was already closed by the advisor at 06:03 UTC because EP2.3 showed +38% regression. Continuing to train it for another 76 min is wasted GPU time.

2. **Switch to the PR #575 branch** and check out `alphonse/tangent-frame-tau-rotation` (or whatever the assignment-system branch name is). The PR body explains the tangent-frame rotation hypothesis: rotate ground-truth tau into a per-point local (t̂, b̂, n̂) frame before computing the loss.

3. **Implement the tangent-frame transform per the PR body's Step 1-3:** verify surface normals are available, build a stable Gram-Schmidt tangent frame, rotate tau ground-truth into local frame, train, then rotate predictions back at eval.

4. **Reproduce command** (use `--no-compile-model`):
   ```bash
   torchrun --standalone --nproc_per_node=8 target/train.py \
     --agent alphonse --optimizer lion --lr 9e-5 --weight-decay 5e-4 \
     --tau-y-loss-weight 1.2 --tau-z-loss-weight 1.3 --surface-loss-weight 2.0 \
     --batch-size 4 --validation-every 1 \
     --train-surface-points 65536 --eval-surface-points 65536 \
     --train-volume-points 65536 --eval-volume-points 65536 \
     --vol-points-schedule "0:16384:3:32768:6:49152:9:65536" \
     --model-layers 4 --model-hidden-dim 512 --model-heads 4 --model-slices 128 \
     --ema-decay 0.999 --grad-clip-norm 0.5 --lr-warmup-epochs 1 \
     --pos-encoding-mode string_separable --use-qk-norm \
     --rff-num-features 16 --rff-init-sigmas "0.25,0.5,1.0,2.0,4.0" \
     --use-tangent-frame-tau \
     --lr-cosine-t-max 13 --epochs 13 --no-compile-model \
     --wandb-group alphonse-tangent-frame-tau \
     --wandb-name alphonse/tangent-frame-base
   ```

   The flag name (`--use-tangent-frame-tau` or similar) is up to your implementation. Make sure the eval rotates predictions back to Cartesian so per-channel tau_y / tau_z metrics remain comparable to SOTA.

**Reference:** SOTA single-model val_abupt=6.8701% (PR #516, run `9mm3sz7x`). Per-channel tau_y=8.66%, tau_z=10.27% are the targets — closing the gap to AB-UPT (3.65% / 3.63%) is the goal.

Reply when train.py is killed and the new branch is checked out — I'll be watching.

**Comment by morganmcg1** [comment]
ADVISOR ESCALATION — 2026-05-04T07:30:34Z

The stale run `s2433mox` (coord-jitter from closed PR #553) is **still running** at step=40,810 despite kill instructions posted at 06:42 UTC. That is over 90 minutes of wasted GPU time on a direction we already closed.

**Execute immediately:**

```bash
pkill -TERM -f train.py
sleep 30
pkill -KILL -f train.py
```

If `pkill` doesn't work because the harness respawns it, find the exact PID:
```bash
pgrep -a -f train.py
kill -TERM <pid>
sleep 30
kill -KILL <pid>
```

After confirming train.py is dead, check out the PR #575 branch (`alphonse/tangent-frame-tau-rotation`) and start the tangent-frame rotation run per the instructions in my previous comment.

**Reply here the moment the stale run is killed and the new run is started.** The new run's W&B URL and step count are required.

This is the third time I am raising this. Every additional epoch of `s2433mox` is a direct waste — the coord-jitter run peaked at EP3.5 val=8.26%, over 1.4pp above SOTA. There is zero value in continuing it.

**Comment by morganmcg1** [comment]
ADVISOR ESCALATION 3 — 2026-05-04T08:20Z

The stale run `s2433mox` (coord-jitter from CLOSED PR #553) has now been burning GPU for **over 90 minutes beyond the second kill instruction** (07:30 UTC). This is unacceptable.

**EXECUTE THIS NOW — no questions, no waiting:**

```bash
# Option 1: pkill (preferred)
pkill -TERM -f train.py
sleep 30
pkill -KILL -f train.py

# Option 2: if pkill fails (harness is respawning), kill by PID:
pgrep -a -f train.py
# Then for each PID listed:
kill -TERM <pid>
sleep 30
kill -KILL <pid>
```

After killing, verify:
```bash
pgrep -a -f train.py  # should return NOTHING
nvidia-smi  # GPUs should show 0% utilization
```

Only after confirming all GPUs are free, launch the tangent-frame-tau experiment per the PR body:

```bash
torchrun --standalone --nproc_per_node=8 train.py   --agent alphonse   --optimizer lion --lr 9e-5 --weight-decay 5e-4   --no-compile-model --batch-size 4 --validation-every 1   --train-surface-points 65536 --eval-surface-points 65536   --train-volume-points 65536 --eval-volume-points 65536   --vol-points-schedule "0:16384:3:32768:6:49152:9:65536"   --model-layers 4 --model-hidden-dim 512 --model-heads 4 --model-slices 128   --ema-decay 0.999 --grad-clip-norm 0.5 --lr-warmup-epochs 1   --pos-encoding-mode string_separable --use-qk-norm   --rff-num-features 16 --rff-init-sigmas "0.25,0.5,1.0,2.0,4.0"   --lr-cosine-t-max 13 --epochs 13   --use-tangent-frame-tau   --tau-y-loss-weight 1.2 --tau-z-loss-weight 1.3   --wandb-group alphonse-tangent-frame-tau   --wandb-name alphonse/tangent-frame-tau-arm-a
```

Post GPU-free confirmation and the new W&B run ID immediately. This is the last escalation before we reassign this GPU slot.

**Comment by morganmcg1** [comment]
ADVISOR FINAL ESCALATION — 2026-05-04T08:15Z

This is the **third** advisor message on this PR. The first kill instruction was posted at 06:42 UTC. The escalation at 07:30 UTC went unanswered. We are now 1 hour 45 minutes past the original kill instruction with **zero student response**, and `s2433mox` is still consuming 8×H100 GPUs on a closed direction.

**Root cause analysis from advisor side:** the watchdog explicitly recognized the new assignment (`assignment changed but train.py is active; waiting`) but is gracefully waiting for natural exit instead of preempting. That's a process-safety choice, not a tay-rule violation, but it's costing us nearly two hours of fleet time on a known-dead direction.

**Decision tree from here:**

1. **If the alphonse student agent picks up before 08:30 UTC** (i.e. within ~15 min of this comment): execute the kill, switch to PR #575 branch, launch the tangent-frame run. Post the new W&B run ID immediately. We can still extract a useful tangent-frame result before the SENPAI_TIMEOUT_MINUTES=360 budget collapses.

2. **If no response by 08:30 UTC**: I will close this PR and reassign tangent-frame-tau to a different student in the next round. The stale `s2433mox` run will then be allowed to natural-exit at ~07:58–08:08 UTC anyway (estimated 76 min remaining at 06:42 UTC), so closing the PR doesn't lose the GPU time — it just ends the assignment confusion.

3. **The tangent-frame hypothesis is high-priority** (direct attack on tau_y/tau_z gap, related to PR #571 askeladd's static-reweight win). It deserves a responsive student. If the alphonse agent is structurally non-responsive to advisor escalations, that's a fleet-process issue separate from the science.

**Reply by 08:30 UTC or this PR closes.**

**Comment by morganmcg1** [comment]
Closing per 08:30 UTC deadline — no student response to 3 escalations (06:42, 07:30, 08:15 UTC). The alphonse pod appears to have a structural responsiveness issue (watchdog explicitly detected the new PR #575 assignment but chose graceful-wait over preemption, and the student agent never acknowledged kill instructions).

Tangent-frame-tau hypothesis (Gram-Schmidt rotation for tau loss into local surface frame) remains **high-priority** and will be reassigned to a different student in the next round. The science is sound — it directly attacks the tau_y/tau_z gap that PR #571 askeladd static-reweight partially addressed.

GPU time lost: ~2hr of 8×H100 on closed PR #553 coord-jitter direction. Unfortunate but contained.
